lib/puppet/functions/postgresql/postgresql_password.rb

```diff
@@ -22,18 +22,21 @@
     required_param 'Variant[String[1], Integer]', :username
     required_param 'Variant[String[1], Sensitive[String[1]], Integer]', :password
     optional_param 'Boolean', :sensitive
-    optional_param "Optional[Enum['md5', 'scram-sha-256']]", :hash
+    optional_param 'Optional[Postgresql::Pg_password_encryption]', :hash
     optional_param 'Optional[Variant[String[1], Integer]]', :salt
     return_type 'Variant[String, Sensitive[String]]'
   end
 
   def default_impl(username, password, sensitive = false, hash = 'md5', salt = nil)
-    return password if password.is_a?(String) && password.match?(%r{^(md5|SCRAM-SHA-256).+})
-
     password = password.unwrap if password.respond_to?(:unwrap)
-    pass = if hash == 'md5'
+    if password.is_a?(String) && password.match?(%r{^(md5[0-9a-f]{32}$|SCRAM-SHA-256\$)})
+      return Puppet::Pops::Types::PSensitiveType::Sensitive.new(password) if sensitive
+      return password
+    end
+    pass = case hash
+           when 'md5', nil # ensure default value when definded with nil
              "md5#{Digest::MD5.hexdigest(password.to_s + username.to_s)}"
-           else
+           when 'scram-sha-256'
              pg_sha256(password, (salt || username))
            end
```

Comment:
Adding an `else` that raise an exception for the `case` statement will help adding new hashing methods if we need to in the future.

Reply:
Ok, added it but in normal usage it is a kind of useless code since the dispatch already ensures that only known hash types are handed over.